flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java

```diff
@@ -1213,4 +1268,17 @@ <T extends State> T transitionToState(StateFactory<T> targetState) {
     State getState() {
         return state;
     }
+
+    /**
+     * Check for slots that are idle for more than {@link JobManagerOptions#SLOT_IDLE_TIMEOUT} and
+     * release them back to the ResourceManager.
+     */
+    private void checkIdleSlotTimeout() {
+        declarativeSlotPool.releaseIdleSlots(System.currentTimeMillis());
+        getMainThreadExecutor()
+                .schedule(
+                        this::checkIdleSlotTimeout,
```

Comment:
1. Could you explain why this method is necessary -don't we release the excessive slots in `CreatingExecutionGraph`? What is the scenario?

2.  Should the scheduled task be cancelled, e.g. in case of losing leadership? Otherwise, won't we get as many scheduled tasks as leadership changes?

Reply:
> Could you explain why this method is necessary -don't we release the excessive slots in CreatingExecutionGraph? What is the scenario?

We mark the slots as idle with the current timestamp and still need to keep them around until `current timestamp + slot idle timeout`; it's up to the slot pool to decide when slots should be physically released

Reply:
> Should the scheduled task be cancelled, e.g. in case of losing leadership? Otherwise, won't we get as many scheduled tasks as leadership changes?

I've added another branch to ensure we don't re-schedule the idleness check when the job was suspended.

Reply:
There should only be at most 1 task being scheduled in the current version of the PR.
If it happens to run after the job was suspended it will just exit.